parquet-common/src/main/java/org/apache/parquet/type/Float16.java

```diff
@@ -0,0 +1,307 @@
+  // Negative 0 of type half-precision float.
+  public static final short NEGATIVE_ZERO = (short) 0x8000;
+  // Byte array in little endian for positive 0 of type half-precision float.
+  public static final byte[] POSITIVE_ZERO_BYTES_LITTLE_ENDIAN = Float16.toBytesLittleEndian(Float16.POSITIVE_ZERO);
+  // Byte array in little endian for negative 0 of type half-precision float.
+  public static final byte[] NEGATIVE_ZERO_BYTES_LITTLE_ENDIAN = Float16.toBytesLittleEndian(Float16.NEGATIVE_ZERO);
+  // A Not-a-Number representation of a half-precision float.
+  public static final short NaN = (short) 0x7e00;
+  // Positive infinity of type half-precision float.
+  static final short POSITIVE_INFINITY = (short) 0x7c00;
+  // Negative infinity of type half-precision float.
+  static final short NEGATIVE_INFINITY = (short) 0xfc00;
+
+  // The bitmask to and a number with to obtain the sign bit.
+  private static final int SIGN_MASK                 = 0x8000;
+  // The offset to shift by to obtain the exponent bits.
+  private static final int EXPONENT_SHIFT            = 10;
+  // The bitmask to and a number shifted by EXPONENT_SHIFT right, to obtain exponent bits.
+  private static final int SHIFTED_EXPONENT_MASK     = 0x1f;
+  // The bitmask to and a number with to obtain significand bits.
+  private static final int SIGNIFICAND_MASK          = 0x3ff;
+  // The offset of the exponent from the actual value.
+  private static final int EXPONENT_BIAS             = 15;
+  // The offset to shift by to obtain the sign bit.
+  private static final int SIGN_SHIFT                = 15;
+  // The bitmask to AND with to obtain exponent and significand bits.
+  private static final int EXPONENT_SIGNIFICAND_MASK = 0x7fff;
+
+  private static final int FP32_SIGN_SHIFT            = 31;
+  private static final int FP32_EXPONENT_SHIFT        = 23;
+  private static final int FP32_SHIFTED_EXPONENT_MASK = 0xff;
+  private static final int FP32_SIGNIFICAND_MASK      = 0x7fffff;
+  private static final int FP32_EXPONENT_BIAS         = 127;
+  private static final int FP32_QNAN_MASK             = 0x400000;
+  private static final int FP32_DENORMAL_MAGIC = 126 << 23;
+  private static final float FP32_DENORMAL_FLOAT = Float.intBitsToFloat(FP32_DENORMAL_MAGIC);
+
+  /**
+   * Converts the specified half-precision float value into a
+   * single-precision float value. The following special cases are handled:
+   * If the input is NaN, the returned value is Float NaN.
+   * If the input is POSITIVE_INFINITY or NEGATIVE_INFINITY, the returned value is respectively
+   *   Float POSITIVE_INFINITY or Float NEGATIVE_INFINITY.
+   * If the input is 0 (positive or negative), the returned value is +/-0.0f.
+   * Otherwise, the returned value is a normalized single-precision float value.
+   *
+   * @param h The half-precision float value to convert to single-precision
+   * @return A normalized single-precision float value
+   */
+  public static float toFloat(short h) {
```

Comment:
Should we use `Binary` instead of `short` as the parameter type here and below? We can call `Binary.get2BytesLittleEndian()` internally. The point is that `Binary` is the actual type holding the float16 values while `short` type is just an intermediate type used internally in the `Float16` class. WDYT? @zhangjiashen @gszadovszky

Reply:
Of course the methods with short variants can be kept for testing purposes.

Reply:
Of course the methods with short variants can be kept for testing purposes.

Reply:
The question is what do we want to use this class for. Inside parquet-mr we need this functionality to add support for `FLOAT16` into `PrimitiveComparator` and `PrimitiveStringifier`. For that we only need a utility class in the same module/package and don't need public methods at all.
If we want to support our API users than we might need public methods that can read/write from/to 2-bytes long `Binary` values. But that would be the first logical type that we would add support for our users.

Reply:
I agree with you @gszadovszky. As I have stated in a comment elsewhere, Float16 is better to be used internally. 

To achieve this, we may need to move Float16.java to package `parquet-column/src/main/java/org/apache/parquet/schema` and remove all `public` identifiers from it so `PrimitiveComparator` and `PrimitiveStringifier` can access it without problem. Unit test under same package will not have problem either and we can always wrap them in a public test class to make it available in the other test modules. 

The only thing is that current implementation also needs `Float16` to be visible to `Float16Builder.build()`. Could you do some refactoring to avoid this? @zhangjiashen. This patch overall looks good to me (except those unresolved comments), thanks!

Reply:
It makes sense to use Binary instead of short and I moved Float16 to **parquet-column/src/main/java/org/apache/parquet/schema** and refactored corresponding code as well, please help take a look?